mule-extensions-api/src/main/java/org/mule/runtime/extension/api/component/value/ValueDeclarer.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.extension.api.component.value;
+
+import org.mule.api.annotation.NoImplement;
+import org.mule.runtime.api.connection.ConnectionProvider;
+
+/**
+ * Describes a value to be created.
+ *
+ * @since 1.5.0
+ */
+@NoImplement
+public interface ValueDeclarer {
+
+  /**
+   * Declares that the described value represents a Map value and returns a declarer to describe it.
+   *
+   * @return the declarer to be used to declare the map value
+   */
+  MapValueDeclarer asMapValue();
+
+  /**
+   * Declares that the described value represents an Object value and returns a declarer to describe it.
+   *
+   * @return the declarer to be used to declare the object value
+   */
+  ObjectValueDeclarer asObjectValue();
```

Comment:
this pattern feels kind of weird an inconsistent. For instance, when I want to add an item to an array, I get a `Consumer<ValueDeclarer>` , but here I get the declarer as a return type, which breaks the daisy chain and becomes a lot more `builder-ish`.

Have you consider pros/cons of instead doing this as `ValueDeclarer objectValue(Consumer<ObjectValueDeclarer>)` ?

Reply:
If I understood correctly, I would actually change it to `void objectValue(Consumer<ObjectValueDeclarer>)` , since I would be passing a consumer, I no longer need the value declarer back.

This is an example of how the declaration of an object inside another object would look with both proposals

Current state of the PR:

```
valueDeclarer -> valueDeclarer.asObjectValue()
    .withField(POJO_CONNECTION_DESCRIPTION_FIELD_NAME, POJO_CONNECTION_DESCRIPTION_FIELD_VALUE)
    .withField(POJO_CONNECTION_TYPE_FIELD_NAME, POJO_CONNECTION_TYPE_FIELD_VALUE.name())
    .withField(POJO_CONNECTION_IMPORTED_POJO_FIELD_NAME, 
	importedPojoValueDeclarer -> importedPojoValueDeclarer.asObjectValue()
        	.withField(IMPORTED_POJO_ID_FIELD_NAME, IMPORTED_POJO_ID_FIELD_VALUE)
        	.withField(IMPORTED_POJO_BOOLEAN_FIELD_NAME, String.valueOf(IMPORTED_POJO_BOOLEAN_FIELD_VALUE)));
```

The change I think you are proposing : `Using void objectValue(Consumer<ObjectValueDeclarer>)`

```
valueDeclarer -> valueDeclarer.objectValue(
	objectValueDeclarer -> objectValueDeclarer
	    .withField(POJO_CONNECTION_DESCRIPTION_FIELD_NAME, POJO_CONNECTION_DESCRIPTION_FIELD_VALUE)
	    .withField(POJO_CONNECTION_TYPE_FIELD_NAME, POJO_CONNECTION_TYPE_FIELD_VALUE.name())
	    .withField(POJO_CONNECTION_IMPORTED_POJO_FIELD_NAME,
       	        innerValueDeclarer -> innerValueDeclarer.objectValue(
			innerObjectValueDeclarer -> innerObjectValueDeclarer
                  		.withField(IMPORTED_POJO_ID_FIELD_NAME, IMPORTED_POJO_ID_FIELD_VALUE)
                   		.withField(IMPORTED_POJO_BOOLEAN_FIELD_NAME, String.valueOf(IMPORTED_POJO_BOOLEAN_FIELD_VALUE)))));
```

I do get that it is more consistent to also make the method receive the Consumer<ObjectValueDeclarer> instead of returning the ObjectValueDeclarer directly. The pro that I like from the current state, is that you avoid the “double lambda" that you see in the second example every time you need to describe an object or an array and I think is may make its usage easier to read.

` valueDeclarer -> valueDeclarer.objectValue( objectValueDeclarer -> objectValueDeclarer.withField(…)… )   `                       

Vs

`valueDeclarer -> valueDeclarer.asObjectValue().withField()…`

What do you think?

Reply:
the double lambda I think it's actually an advantage as it makes composition cases easier. Don't forget that in your example you're statically defining things, but as we learned from the ExtensionDeclarer experience, in real life you end up having intermediate logic and reusable methods for common structures. Also, your example would be much more readable if you used better names for the consumer var. For instance `objectValueDeclarer -> obj` or `innerObjectValueDeclarer -> inner`

Reply:
Agreed on using void instead of daisy